modules/raft-api/src/main/java/org/apache/ignite/internal/raft/Command.java

```diff
@@ -17,10 +17,27 @@
 
 package org.apache.ignite.internal.raft;
 
+import org.apache.ignite.internal.hlc.HybridTimestamp;
 import org.apache.ignite.internal.network.NetworkMessage;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * A marker interface for replication group command.
  */
 public interface Command extends NetworkMessage {
+    /**
+     * This is called before a command is submitted to replication pipeline.
+     *
+     * @param safeTs Safe timestamp.
+     */
+    default void patch(HybridTimestamp safeTs) {}
```

Comment:
The concept of patching the command isn't actually good. Meaning that command becomes mutable and thus not-thread safe. Let's say on the replica side, within one of the replica's threads ,UpdateAllCommand was created a bit later, exactly that command was patched with some other in-raft thread. Without extra synchronization, replica won't see proposed safeTime. Basically, it's quite fragile.

Reply:
We already have this concept. See patchCommandBeforeApply. This looks really ugly.
My PR actually makes this concept clean. 
Patching is hidden on replication layer and not visible outside. Message may or may not be patched, depending on protocol requirements.
No thread safety issues in the current implementation.

Reply:
> We already have this concept. 
Well, thats true. However you are currently reworking the logic. Why not to remove 'ugly' part of the code.

> No known to me thread safety issues exist in the PR.
Reading safeTime anyware besides raft linearised logic isn't thread safe. Meaning that if i'll read original command.safeTime that was created on the Replica side I may not see patched value. I don't think that currently anyone does that, however it's just fragile. At least you may add a comment to safeTime getter explaining that it's not thread safe.

Reply:
> Why not to remove 'ugly' part of the code.

I did exactly that. The next step is to remove beforeApply handler from metastore.
> Reading safeTime anyware besides raft linearised logic isn't thread safe

I don't get you. safeTime is no longer assigned on primary replica. It's assinged in a fully thread safe way inside replication layer. All safeTs usages are thread safe in the PR.

Reply:
> I don't get you. safeTime is no longer assigned on primary replica. It's assinged in a fully thread safe way on entering replication layer. All safeTs usages are thread safe in the PR.

I mean what if someone will try to 

- create UpdateCommand cmd in PartitionReplicaListener, send it to Raft whehere it'll be patched.
- will call cmd.safeTime() on the PartitionReplicaListener side. What will he see?

Reply:
I think I got you.
Reading safeTime after patching the class in another thread is not thread safe in the current implementation. This is mutability consequences.
Javadoc on safeTime already mentions special processing for this field:
/** The Safe timestamp. A value for this field is auto generated and should not be set manually. */
So it should never be read back.
This however can be fixed by cloning the command before submitting it to raft, but IMO doesn't give any profit.

Reply:
I'd also mention in `safeTime();` javadoc that it's not thread-safe, and that proper visibility is only guaranteed inside raft.

Reply:
🆗